go/apps/mixnet/README.md

```diff
@@ -0,0 +1,168 @@
+To send a message to a server, the client sends the message to the proxy
+which divides the message into cells and sends them to the router. The
+router waits until it has received all the cells, then assembles them into
+the original message. Once the router has queued messages from enough senders, it
+transmits the message to the server. It then waits for a reply from the server,
+divides the reply into cells, and queues the cells to be transmitted back to
+the proxy. The proxy waits until it has received all the reply cells, assembles
+them into the complete message, and sends it back to the client.
+
+The router maintains two datastructures: the _sendQueue_ for sender to recipient
+traffic, and the _replyQueue_ for recipient to sender traffic. They are
+functionally equivalent (see `queue.go`); they receive cells from proxies (or other routers)
+and replies from recipients, and they send cells to proxies (or other routers)
+and messages to recipients. The _batchSize_ specifies how many messages/cells
+are waiting in the queue from distinct peers before transmitting
+them. When its time to transmit, a connection to the destination is established
+(if it hasn't been established already) and the messages are sent in a random
+order.
+
+A cell is either a chunk of a message or a _directive_. A directive contains
+instructions for a router or proxy and are implemented as protocol buffers
+(`mixnet.proto`):
+
+ * ERROR: something went wrong.
+ * CREATE: the proxy instructs the router to create a circuit over the mixnet
+           to a destination address. The response is either ERROR or CREATED.
+ * CREATED: the router informs the proxy that the circuit was created.
+ * DESTROY: the proxy instructs the router to destroy the circuit.
+
+Note that when a circuit is "created", all that happens is the router is
+informed of the destination server. The connection is established when the
+router is ready to dequeue; otherwise, the TCP handshake would allow the
+adversary to correlate the wrapped CREATE directive with the TCP handshake.
+
+The router (`router.go`) is a Tao-delegated program and will only run when launched in a
+Tao environment. The proxy (`proxy.go`) is not launched in the Tao and is expected to run
+locally on the client's machine. It is assumed that the client has a copy
+of the root public policy key. When the proxy dials the router, the router
+attests to its identity, and the attestation is verified by the proxy using
+the policiy key; if verification fails, the proxy exits. Both the proxy and
+router communicate with the TaoCA to obtain a copy of the policy.
+
+The proxy exposes the SOCKS protocol to clients (`socks5.go`), a widely used protocol
+for proxying internet traffic specified in RFC 1928 [3]. Our proxy only
+partially implements the server role; see `SocksListener.Accept()`
+for details. For example, it only allows the client to specify
+IPv4 addresses; hostnames need to be allowed so that mixnet can proxy DNS
```

Comment:
Huh, I suppose it could. When you type in "google.com", the browser first hits a DNS server (e.g. 8.8.8.8) to resolve the IP address for hostname "google.com". Then it sends a request to the IP address. I guess there's no reason to pass the hostname through the mixnet; just proxy the DNS request as well?